### PR TITLE
testcase/filesystem: Improve filesystem vfs testcase

### PR DESCRIPTION
**Pull request by KimDongEon**
1. Remove dependency of vfs tc
2. Separation Positive Case and Negative Case
3. Some TC bug fix

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [2d8dba09153c899d21e1aed3f1274be941a5be3a] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [2d8dba09153c899d21e1aed3f1274be941a5be3a] - Code Rule Check OK.

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 1]
```diff
@@ -273,11 +365,27 @@ static void tc_fs_vfs_mount(void)
...
 	TC_ASSERT_EQ("umount", ret, OK);
+
+	TC_SUCCESS_RESULT();
+}
+
+static void tc_fs_vfs_umount_invalid_path_n(void)
+{
+	int ret;
+
+	tc_fs_vfs_init_unmount();
```
Comment (sangwon03) [inline apps/examples/testcase/le_tc/filesystem/fs_main.c]: is this correct to use 'tc_fs_vfs_init_unmount'?

Comment (KimDongEon) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: This tc is the case for unmount without any unmount path.

So to make sure that there is no mount path at initialization
The call "tc_fs_vfs_init_unmount ()" seems appropriate.

However, the name of this tc is more appropriate "tc_fs_vfs_umount_noexist_path_n".

I will modify tc name "tc_fs_vfs_umount_noexist_path_n"

**Comment by seinfra** [comment]
Target : [bb26de1e5968a2f1113cc5652ce64c6f325f832f] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [bb26de1e5968a2f1113cc5652ce64c6f325f832f] - Code Rule Check OK.

**Comment by KimDongEon** [comment]
@pillip8282
I reflected your comment.

**Comment by seinfra** [comment]
Target : [e5af1a46df6e49f54385ad54fc209e0fe1fae6a7] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [e5af1a46df6e49f54385ad54fc209e0fe1fae6a7] - Code Rule Check Result:
apps/examples/testcase/le_tc/filesystem/fs_main.c:1561: ERROR: [SPC_M_SEP] space prohibited between function name and open parenthesis '('

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 2]
```diff
@@ -289,19 +421,39 @@ static void tc_fs_vfs_umount(void)
...
+
+	/* Init */
+	vfs_mount();
+
+	/* Testcase */
 	fd = open(VFS_FILE_PATH, O_WROK | O_CREAT);
-	TC_ASSERT_GEQ("open", fd, 0);
+	TC_ASSERT_GEQ_CLEANUP("open", fd, 0, vfs_unmount());
+
+	/* Deinit */
```
Comment (drashti304) [inline apps/examples/testcase/le_tc/filesystem/fs_main.c]: This line should come below before vfs_umount();

Comment (KimDongEon) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: This code is to execute deinit if open() fails.

Can you explain why this line should come before vfs_umount();?

**Comment by seinfra** [comment]
Target : [4b1e1aa7cee75d968190e481871cb2d1cbcb9f5b] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [4b1e1aa7cee75d968190e481871cb2d1cbcb9f5b] - Code Rule Check OK.

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 3]
```diff
@@ -496,39 +765,56 @@ static void tc_fs_vfs_dup2(void)
...
 	ret = read(fd2, buf, len);
+
+	TC_ASSERT_GT_CLEANUP("read", ret, 0, close(fd2));
+	TC_ASSERT_EQ_CLEANUP("read", strcmp(buf, VFS_TEST_CONTENTS_3), 0, close(fd2));
+
+	/* Deinit */
 	close(fd2);
-	TC_ASSERT_GT("read", ret, 0);
-	TC_ASSERT_EQ("read", strcmp(buf, VFS_TEST_CONTENTS_3), 0);
+	vfs_unmount();
```
Comment (chanijjani) [inline apps/examples/testcase/le_tc/filesystem/fs_main.c]: There are too many "vfs_unmount()" at the end of TCs, so I'll stop commeting about this here.
Please modify remaining vfs_unmount() if needed.

**Comment by chanijjani** [comment]
@KimDongEon Is there any reason to change TC_ASSERT_XX into TC_ASSER_XX_CLEANUP?

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 4]
```diff
@@ -430,49 +684,64 @@ static void tc_fs_vfs_dup(void)
...
+	TC_SUCCESS_RESULT();
+}
 
+static void tc_fs_vfs_dup_invalid_fd_n(void)
+{
+	int fd;
 	/* Nagative case with invalid argument, invalid fd. It will return ERROR */
+
+	/* Testcase */
 #if CONFIG_NFILE_DESCRIPTORS > 0
```
Comment (junmin-kim) [inline apps/examples/testcase/le_tc/filesystem/fs_main.c]: CONFIG_FILE_DESCRIPTORS always >= 0 isn't it?

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 5]
```diff
@@ -705,28 +1069,68 @@ static void tc_fs_vfs_pread(void)
...
-		ret = mkdir(filename, 0777);
-		TC_ASSERT_EQ("mkdir", ret, OK);
-	}
 	/*creating an empty folder */
 	ret = mkdir(DEV_EMPTY_FOLDER_PATH, 0777);
-	TC_ASSERT_EQ("mkdir", ret, OK);
+	TC_ASSERT_EQ_CLEANUP("mkdir", ret, OK, vfs_unmount());
+
+	/* Deinit */
+	vfs_unmount();
```
[the remark this answers is not in the document]
Comment (KimDongEon) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: To do a complete deinit() you need to run a function that cleans up before the end of the function, regardless of the ret result.

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 6]
```diff
@@ -705,28 +1069,68 @@ static void tc_fs_vfs_pread(void)
...
+	/* Init */
+	vfs_mount();
+	vfs_mkdir(VFS_FOLDER_PATH);
+
+	/* Testcase */
+	ret = mkdir(VFS_FOLDER_PATH, 0777);
+	TC_ASSERT_EQ_CLEANUP("mkdir", ret, ERROR, vfs_unmount());
+
+	/* deinit */
+	vfs_unmount();
```
[the remark this answers is not in the document]
Comment (KimDongEon) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: To do a complete deinit() you need to run a function that cleans up before the end of the function, regardless of the ret result.

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 7]
```diff
@@ -705,28 +1069,68 @@ static void tc_fs_vfs_pread(void)
...
+	for (i = 0; i < CONFIG_NAME_MAX + 1; i++) {
+		sprintf(filename, "%s%c", filename, 'a');
+	}
+
+	/* Testcase */
+	ret = mkdir(filename, 0777);
+	TC_ASSERT_EQ_CLEANUP("mkdir", ret, ERROR, free(filename); vfs_unmount());
+
+	/* Deinit */
+	free(filename);
```
[the remark this answers is not in the document]
Comment (KimDongEon) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: As above, free() must be called before the function exits to complete memory cleanup.

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 8]
```diff
@@ -705,28 +1069,68 @@ static void tc_fs_vfs_pread(void)
...
+	vfs_unmount();
```
[the remark this answers is not in the document]
Comment (KimDongEon) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: To do a complete deinit() you need to run a function that cleans up before the end of the function, regardless of the ret result.

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 9]
```diff
@@ -250,21 +334,46 @@ static int make_long_file(void)
...
 
+	/*search mountpoint operation*/
 #if !defined(CONFIG_DISABLE_ENVIRON) && !defined(CONFIG_BUILD_PROTECTED)
 	ret = mount_show();
-	TC_ASSERT_EQ("mount_show", ret, OK);
+	TC_ASSERT_EQ_CLEANUP("mount_show", ret, OK, vfs_unmount());
 #endif
+
+	/* Deinit */
+	vfs_unmount();
```
[the remark this answers is not in the document]
Comment (KimDongEon) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: To do a complete deinit() you need to run a function that cleans up before the end of the function, regardless of the ret result.

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 10]
```diff
@@ -366,28 +568,59 @@ static void tc_fs_vfs_write(void)
...
+	/* Testcase */
 	ret = read(fd, buf, sizeof(buf));
+	TC_ASSERT_GEQ_CLEANUP("read", ret, 0, close(fd); vfs_unmount());
+	TC_ASSERT_EQ_CLEANUP("read", strcmp(buf, VFS_TEST_CONTENTS_1), 0, close(fd); vfs_unmount());
+
+	/* Deinit */
 	close(fd);
-	TC_ASSERT_GEQ("read", ret, 0);
-	TC_ASSERT_EQ("read", strcmp(buf, VFS_TEST_CONTENTS_1), 0);
+	vfs_unmount();
```
[the remark this answers is not in the document]
Comment (KimDongEon) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: To do a complete deinit() you need to run a function that cleans up before the end of the function, regardless of the ret result.

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 11]
```diff
@@ -366,28 +568,59 @@ static void tc_fs_vfs_write(void)
...
 	memset(buf, 0, sizeof(buf));
+
+	/* Testcase */
 	ret = read(fd, buf, sizeof(buf));
+	TC_ASSERT_LT_CLEANUP("read", ret, 0, close(fd); vfs_unmount());
+
+	/* Deinit */
 	close(fd);
-	TC_ASSERT_LT("read", ret, 0);
+	vfs_unmount();
```
[the remark this answers is not in the document]
Comment (KimDongEon) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: To do a complete deinit() you need to run a function that cleans up before the end of the function, regardless of the ret result.

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 12]
```diff
@@ -402,13 +635,34 @@ static void tc_fs_vfs_read(void)
...
 	fd = open(VFS_FILE_PATH, O_RDONLY);
-	TC_ASSERT_GEQ("open", fd, 0);
+	TC_ASSERT_GEQ_CLEANUP("open", fd, 0, vfs_unmount());
+
+	/* Testcase */
+	ret = close(fd);
+	TC_ASSERT_GEQ_CLEANUP("close", ret, 0, close(fd); vfs_unmount());
+
+	/* Deinit */
+	vfs_unmount();
```
[the remark this answers is not in the document]
Comment (KimDongEon) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: To do a complete deinit() you need to run a function that cleans up before the end of the function, regardless of the ret result.

File: apps/examples/testcase/le_tc/filesystem/fs_main.c (hunk 3 above)
Comment (KimDongEon) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: As in the case above, at the point where the function terminates, it is terminated after the deinit operation.

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 13]
```diff
@@ -402,13 +635,34 @@ static void tc_fs_vfs_read(void)
...
+
+	/* Init */
+	vfs_mount();
 	fd = open(VFS_FILE_PATH, O_RDONLY);
-	TC_ASSERT_GEQ("open", fd, 0);
+	TC_ASSERT_GEQ_CLEANUP("open", fd, 0, vfs_unmount());
+
+	/* Testcase */
+	ret = close(fd);
+	TC_ASSERT_GEQ_CLEANUP("close", ret, 0, close(fd); vfs_unmount());
```
[the remark this answers is not in the document]
Comment (KimDongEon) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: After the close () call failed, it is confirmed that there is no reason to succeed even if the close () is executed again in XX_CLEANUP.

TC_ASSERT_GEQ_CLEANUP("close", ret, 0, vfs_unmount());
I'll modify that code.

**Comment by seinfra** [comment]
Target : [f3a0902366646913a18371026d19a3e9ee0bdb6b] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [f3a0902366646913a18371026d19a3e9ee0bdb6b] - Code Rule Check Result:
apps/examples/testcase/le_tc/filesystem/fs_main.c:1561: ERROR: [SPC_M_SEP] space prohibited between function name and open parenthesis '('

**Comment by seinfra** [comment]
Target : [c4030bd88b4cade83e461049d20b81366daa098b] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [c4030bd88b4cade83e461049d20b81366daa098b] - Code Rule Check (C++) OK.

**Comment by chanijjani** [comment]
> @chanijjani TC_ASSERT_XX exits the function without deiniting at that point if the result is not as expected.
   --> Yes.
> However, if TC_ASSER_XX_CLEANUP is specified a function to perform deinit, it will terminate after executing the function.
   --> But you added deinit functions at the end of all TCs, then TC_ASSERT_XX is also okay, isn't it?
> Init and deinit were performed to remove dependency of all testcases.
   --> Okay.

**Comment by KimDongEon** [comment]
@chanijjani
1. All TCs are success in both real boards and QEMU (including vfs_readdir) after apply this PR,
--> In QEMU, artik053 and imxrt1020, all vfs tc have passed. However, non-vfs tc fails because it relies on vfs tc before being changed by PR. This PR is one of the processes that removes the dependency of all file system test case. So, in the next PR, I will solve the rest of the testcase that fails with dependency issues.

--> "vfs_readdir" passed all of the above.
Is there a reason to check vfs_readdir separately?

2. The TCT coverage of file system increases at least a little.
--> Do you mean CTC coverage?
--> The purpose of this PR is to improve the problem with each test case dependency and to separate the positive and negative cases.
--> If you need to measure CTC coverage, we recommend that you try after all test cases have been improved. The reason is that there is still a follow-up PR.

**Pull request merged**